prusti-interface/src/specs/typed.rs

```diff
@@ -32,79 +45,232 @@ impl SpecificationSet {
+                _ => {}
+            }
+        }
+
+        if matches!(other, Refined(_, _)) {
+            panic!("Can not refine with a refined item");
+        }
+
+        let refined = if self.is_empty() && !other.is_empty() {
+            match other.clone() {
+                Inherent(val) => Inherited(val),
+                Inherited(val) => Inherited(val),
+                Empty | Refined(_, _) => unreachable!()
+            }
+        } else if !self.is_empty() && !other.is_empty() {
+            let (_, self_val) = self.get().unwrap();
+            let (_, other_val) = other.get().unwrap();
+            Refined(other_val.clone(), self_val.clone())
+        } else {
+            self
+        };
+
+        trace!("\t -> {:?}", refined);
+        refined
+    }
+}
+
+impl Refinable for ProcedureSpecification {
+    fn refine(self, other: &Self) -> Self {
+        ProcedureSpecification {
+            pres: self.pres.refine(&other.pres),
+            posts: self.posts.refine(&other.posts),
+            pledges: self.pledges.refine(&other.pledges),
+            predicate_body: self.predicate_body.refine(&other.predicate_body),
+            pure: self.pure.refine(&other.pure),
+            trusted: self.trusted.refine(&other.trusted),
+        }
+    }
+}
+
+impl Refinable for SpecificationSet {
+    fn refine(self, other: &Self) -> Self {
+        if self.is_procedure() && other.is_procedure() {
+            let self_proc = self.into_procedure();
+            let other_proc = other.expect_procedure();
+            let refined = self_proc.refine(other_proc);
+            return SpecificationSet::Procedure(refined);
+        }
+        self
+    }
+}
+
+
 #[derive(Debug, Clone)]
 pub struct ProcedureSpecification {
-    pub pres: Vec<LocalDefId>,
-    pub posts: Vec<LocalDefId>,
-    pub pledges: Vec<Pledge>,
-    pub predicate_body: Option<LocalDefId>,
-    pub pure: bool,
-    pub trusted: bool,
+    pub pres: SpecificationItem<Vec<LocalDefId>>,
+    pub posts: SpecificationItem<Vec<LocalDefId>>,
+    pub pledges: SpecificationItem<Vec<Pledge>>,
+    pub predicate_body: SpecificationItem<LocalDefId>,
+    pub pure: SpecificationItem<bool>,
+    pub trusted: SpecificationItem<bool>,
 }
 
 impl ProcedureSpecification {
     pub fn empty() -> Self {
         ProcedureSpecification {
-            pres: vec![],
-            posts: vec![],
-            pledges: vec![],
-            predicate_body: None,
-            pure: false,
-            trusted: false,
-        }
-    }
-    /// Trait implementation method refinement
-    /// Choosing alternative C as discussed in
-    /// https://ethz.ch/content/dam/ethz/special-interest/infk/chair-program-method/pm/documents/Education/Theses/Matthias_Erdin_MA_report.pdf
-    /// pp 19-23
-    ///
-    /// In other words, any pre-/post-condition provided by `other` will overwrite any provided by
-    /// `self`.
```

Comment:
Can we keep this comment somewhere nearby?

Reply:
I think the refactoring obfuscated the "refinement strategy" completely. The strategy is deeply hidden in `SpecItemIter`, which I am not super happy about in retrospective.
I think the we could add methods to `ProcedureSpecification` to get the `LocalDefId`s of pre-/postcondition. There it will be clearer what what actual specification will be encoded.

Reply:
For example, to get the preconditions, we would have the following function:
```rust
    pub fn get_preconditions(&self) -> Vec<LocalDefId> {
        match self.pres.clone() {
            SpecificationItem::Empty => vec![],
            SpecificationItem::Inherent(items) => items,
            SpecificationItem::Inherited(items) => items,
            SpecificationItem::Refined(_, items) => items,
        }
    }
```
Which makes it pretty explicit that we use the (C) strategy. 
The `clone()` is a bit ugly though and was the reason of the weird `SpecItemIter`. In an ideal world, I would want `get_preconditions` to return a slice, which is (I think) not possible due to the `SpecificationItem::Empty` variant. Technically, we could drop this variant, but I like it (especially when debugging -> one can observe that the user defined did not specify anything).

Reply:
I agree that the documentation needs to be somewhere and ideally, it is also immediately obvious when viewing the code.

I think `mir/specifications` is not the correct place for this doc. My refactoring basically just adds more information to specs (see changes in `prusti-interface/specs/typed.rs`) and `mir/specifications` just provides (lazy) access to these enriched specifications (including computing refinements if needed).
The actual to-be-encoded assertions are accessed in `ProcedureContractGeneric::{functional_precondition, functional_postcondition, pledges}`, i.e. this is the place where we can choose how to handle this additional information.

Reply:
I made the usage of the strategy more explicit by rearranging stuff in `prusti-interface/specs/typed.rs`